builtin/logical/pki/acme_challenges_test.go

```diff
@@ -219,3 +231,473 @@ func TestAcmeValidateDNS01Challenge(t *testing.T) {
+			shouldFail:   true,
+		}
+		alpnTestCases = append(alpnTestCases, newTc)
+	}
+
+	{
+		// Test case: cert without CN
+		// Compute our authorization.
+		checksum := sha256.Sum256([]byte("valid.valid"))
+		authz := base64.RawURLEncoding.EncodeToString(checksum[:])
+
+		// Build a leaf certificate which should pass validation
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		require.NoError(t, err, "failed generating leaf private key")
+		tmpl := &x509.Certificate{
+			Subject:      pkix.Name{},
+			Issuer:       pkix.Name{},
+			KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+			PublicKey:    key.Public(),
+			SerialNumber: big.NewInt(2),
+			DNSNames:     []string{host},
+			ExtraExtensions: []pkix.Extension{
+				{
+					Id:       OIDACMEIdentifier,
+					Critical: true,
+					Value:    []byte(authz),
+				},
+			},
+			BasicConstraintsValid: true,
+			IsCA:                  true,
+		}
+		certBytes, err := x509.CreateCertificate(rand.Reader, tmpl, tmpl, key.Public(), key)
+		require.NoError(t, err, "failed to create leaf certificate")
+		cert, err := x509.ParseCertificate(certBytes)
+		require.NoError(t, err, "failed to parse newly generated leaf certificate")
+
+		newTc := alpnTestCase{
+			name:         "valid certificate; no Subject/Issuer (missing CN)",
+			certificates: []*x509.Certificate{cert},
+			privateKey:   key,
+			protocols:    []string{ALPNProtocol},
+			token:        "valid",
+			thumbprint:   "valid",
+			shouldFail:   false,
+		}
+		alpnTestCases = append(alpnTestCases, newTc)
+	}
+
+	{
+		// Test case: cert without the extension
+		// Build a leaf certificate which should fail validation
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		require.NoError(t, err, "failed generating leaf private key")
+		tmpl := &x509.Certificate{
+			Subject:               pkix.Name{},
+			Issuer:                pkix.Name{},
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+			PublicKey:             key.Public(),
+			SerialNumber:          big.NewInt(1),
+			DNSNames:              []string{host},
+			BasicConstraintsValid: true,
+			IsCA:                  true,
+		}
+		certBytes, err := x509.CreateCertificate(rand.Reader, tmpl, tmpl, key.Public(), key)
+		require.NoError(t, err, "failed to create leaf certificate")
+		cert, err := x509.ParseCertificate(certBytes)
+		require.NoError(t, err, "failed to parse newly generated leaf certificate")
+
+		newTc := alpnTestCase{
+			name:         "missing required acmeIdentifier extension",
+			certificates: []*x509.Certificate{cert},
+			privateKey:   key,
+			protocols:    []string{ALPNProtocol},
+			token:        "valid",
+			thumbprint:   "valid",
+			shouldFail:   true,
+		}
+		alpnTestCases = append(alpnTestCases, newTc)
+	}
+
+	{
+		// Test case: root without a leaf
```

Comment:
all of these are self-signed, why should this one fail?

Reply:
This is a random root CA, presented as a server certificate. It lacks the extensions required by the challenge RFC (no DNSSans and no acmeIdentifier extension) because it is a root CAs. Root CAs, while yes, self-signed, should not validate and violate other required verification steps.

See https://datatracker.ietf.org/doc/html/rfc8737#section-3:

> The client prepares for validation by constructing a self-signed certificate that MUST contain an acmeIdentifier extension and a subjectAlternativeName extension [RFC5280]. The subjectAlternativeName extension MUST contain a single dNSName entry where the value is the domain name being validated. The acmeIdentifier extension MUST contain the SHA-256 digest [FIPS180-4] of the key authorization [RFC8555] for the challenge. The acmeIdentifier extension MUST be critical so that the certificate isn't inadvertently used by non-ACME software.